### PR TITLE
use vscode-languageserver and util libraries to implement server-client connection

### PR DESCRIPTION
**Pull request by asiandrummer**
This PR refactors how the server is instantiated and the messages are understood, with using `vscode-languageserver` library that includes helper utilities such as message reader/writer, connection handler, request/notification/response handler and sender, and valid parameters/return types for each message. Using this library helps in many ways, including (but not limited to):
- Hyperlink (definition) feature is LSP-compliant (!!!)
- Error handling is baked in for free
- Types are easier to check
- Almost right-out-of-the-box integration with VSCode (which is coming very soon...)

Also this PR allowed for a huge refactoring for IPC protocol support functions and the message processing functions.

Tested with VSCode/IntelliJ and a pre-requisite for a more improved Nuclide support.

Also, this PR might qualify for our "minor" version bump.

### CODE REVIEW

**Comment by asiandrummer** [comment]
cc @martijnwalraven for visibility

File: package.json [hunk 1]
```diff
@@ -39,17 +39,17 @@
...
-    "graphql-language-service-interface": "0.0.7",
-    "graphql-language-service-server": "0.0.10",
-    "graphql-language-service-utils": "0.0.6",
-    "yargs": "^7.0.1",
-    "graphql-language-service-types": "0.0.12",
     "graphql-language-service-config": "0.0.7",
+    "graphql-language-service-interface": "0.0.7",
     "graphql-language-service-parser": "0.0.6",
-    "fb-watchman": "^2.0.0",
+    "graphql-language-service-server": "file:packages/graphql-language-service-server",
```
Comment (asiandrummer) [inline package.json]: ugh, changing

Comment (wincent) [inline package.json]: Seems wrong.

File: packages/graphql-language-service-server/src/MessageProcessor.js [hunk 2]
```diff
@@ -9,409 +9,205 @@
  */
 
 import type {Diagnostic, Uri} from 'graphql-language-service-types';
+import {
+  CompletionRequest,
+  CompletionList,
+  Location,
+  InitializeResult,
+  PublishDiagnosticsParams,
```
Comment (wincent) [inline packages/graphql-language-service-server/src/MessageProcessor.js]: nit: alpha

File: packages/graphql-language-service-server/src/MessageProcessor.js [hunk 3]
```diff
@@ -9,409 +9,205 @@
...
-          if (cachedDocument) {
-            text = cachedDocument.content.text;
-          }
-        }
-      }
+export async function handleDidOpenOrSaveNotification(
+  params: NotificationMessage,
+): Promise<PublishDiagnosticsParams> {
+  if (!params || !params.textDocument) {
+    // `textDocument` is required.
```
Comment (wincent) [inline packages/graphql-language-service-server/src/MessageProcessor.js]: Redundant comment.

File: packages/graphql-language-service-server/src/MessageProcessor.js [hunk 4]
```diff
@@ -9,409 +9,205 @@
...
-      }
+export async function handleDidChangeNotification(
+  params: NotificationMessage,
+): Promise<PublishDiagnosticsParams> {
+  // For every `textDocument/didChange` event, keep a cache of textDocuments
+  // with version information up-to-date, so that the textDocument contents
+  // may be used during performing language service features,
+  // e.g. autocompletions.
+  if (!params || !params.textDocument || !params.contentChanges) {
+    // `textDocument` and `contentChanges` are required.
```
Comment (wincent) [inline packages/graphql-language-service-server/src/MessageProcessor.js]: Kill the comment.

File: packages/graphql-language-service-server/src/MessageProcessor.js [hunk 5]
```diff
@@ -9,409 +9,205 @@
...
+export async function handleDidCloseNotification(
+  params: NotificationMessage,
 ): Promise<void> {
-  if (message.length === 0) {
-    return;
+  // For every `textDocument/didClose` event, delete the cached entry.
+  // This is to keep a low memory usage && switch the source of truth to
+  // the file on disk.
+  if (!params || !params.textDocument) {
+    // `textDocument` is required.
```
Comment (wincent) [inline packages/graphql-language-service-server/src/MessageProcessor.js]: Kill comment.

File: packages/graphql-language-service-server/src/MessageProcessor.js [hunk 6]
```diff
@@ -9,409 +9,205 @@
...
-    languageService = new GraphQLLanguageService(graphQLCache);
+}
+
+export async function handleInitializeRequest(
+  params: RequestMessage,
+  token: CancellationToken,
+  configDir?: string,
+): Promise<InitializeResult.type> {
+  if (!params || !params.rootPath) {
+    // `rootPath` is required
```
Comment (wincent) [inline packages/graphql-language-service-server/src/MessageProcessor.js]: Kill.

Comment (wincent) [inline packages/graphql-language-service-server/src/MessageProcessor.js]: Also, if `params` is Flow-typed as `RequestMessage`, why do we have to check `!params` herer?

File: packages/graphql-language-service-server/src/MessageProcessor.js [hunk 7]
```diff
@@ -9,409 +9,205 @@
...
-          error: {
-            code: ERROR_CODES.PARSE_ERROR,
-            message: 'Request contains incorrect JSON format',
-          },
-        }),
-      ),
-    );
-    return;
+  if (!serverCapabilities) {
+    throw new Error('GraphQL language server is not initialized.');
```
Comment (wincent) [inline packages/graphql-language-service-server/src/MessageProcessor.js]: Capitalize "Language Server"?

File: packages/graphql-language-service-server/src/MessageProcessor.js [hunk 8]
```diff
@@ -9,409 +9,205 @@
...
+export async function handleCompletionRequest(
+  params: CompletionRequest.type,
+  token: CancellationToken,
+): Promise<CompletionList> {
+  // `textDocument/comletion` event takes advantage of the fact that
+  // `textDocument/didChange` event always fires before, which would have
+  // updated the cache with the query text from the editor.
+  // Treat the computed list always complete.
+  if (!params || !params.textDocument || !params.position) {
+    // `textDocument` is required.
```
Comment (wincent) [inline packages/graphql-language-service-server/src/MessageProcessor.js]: Kill.

File: packages/graphql-language-service-server/src/MessageProcessor.js [hunk 9]
```diff
@@ -9,409 +9,205 @@
...
-      process.stdout.write(JSON.stringify(responseMsg) + '\n');
-      break;
-    default:
-      break;
+export async function handleDefinitionRequest(
+  params: CompletionRequest.type,
+  token: CancellationToken,
+): Promise<Array<Location>> {
+  if (!params || !params.textDocument || !params.position) {
+    // `textDocument` is required.
```
Comment (wincent) [inline packages/graphql-language-service-server/src/MessageProcessor.js]: Kill.

File: packages/graphql-language-service-server/src/MessageProcessor.js [hunk 10]
```diff
@@ -9,409 +9,205 @@
...
+    throw new Error('`textDocument` and `position` arguments are required.');
   }
+  const textDocument = params.textDocument;
+  const pos = params.position;
+
+  const cachedDocument = getCachedDocument(textDocument.uri);
+  if (!cachedDocument) {
+    // `cachedDocument` is required.
```
Comment (wincent) [inline packages/graphql-language-service-server/src/MessageProcessor.js]: Kill.

File: packages/graphql-language-service-server/src/startServer.js [hunk 11]
```diff
@@ -11,117 +11,147 @@
...
 import {
-  processStreamMessage,
-  processIPCRequestMessage,
-  processIPCNotificationMessage,
+  handleDidChangeNotification,
+  handleDidCloseNotification,
+  handleDidOpenOrSaveNotification,
+  handleInitializeRequest,
+  handleCompletionRequest,
+  handleDefinitionRequest,
```
Comment (wincent) [inline packages/graphql-language-service-server/src/startServer.js]: nit: alpha

File: packages/graphql-language-service-server/src/startServer.js [hunk 12]
```diff
@@ -11,117 +11,147 @@
...
 
 import {
   IPCMessageReader,
   IPCMessageWriter,
   SocketMessageReader,
   SocketMessageWriter,
+  StreamMessageReader,
+  StreamMessageWriter,
+  MessageConnection,
+  createMessageConnection,
```
Comment (wincent) [inline packages/graphql-language-service-server/src/startServer.js]: nit: alpha

File: packages/graphql-language-service-server/src/startServer.js [hunk 13]
```diff
@@ -11,117 +11,147 @@
...
+  DidOpenTextDocumentNotification,
+  DidSaveTextDocumentNotification,
+  DidChangeTextDocumentNotification,
+  DidCloseTextDocumentNotification,
+  ExitNotification,
+  InitializeRequest,
+  CompletionRequest,
+  CompletionResolveRequest,
+  DefinitionRequest,
+  PublishDiagnosticsNotification,
```
Comment (wincent) [inline packages/graphql-language-service-server/src/startServer.js]: alpha

File: packages/graphql-language-service-server/src/startServer.js [hunk 14]
```diff
@@ -11,117 +11,147 @@
...
   method?: string,
   configDir?: string,
 };
 
 export default (async function startServer(options?: Options): Promise<void> {
-  if (!options || !options.configDir) {
+  if (!options) {
+    process.stderr.write('At least --configDir option is required!');
+    process.exit(1);
+    return;
```
Comment (wincent) [inline packages/graphql-language-service-server/src/startServer.js]: These `return` statements are for Flow? If so, might be worth asking in the FB Flow group why Flow doesn't know that `process.exit` terminates execution.

File: packages/graphql-language-service-server/src/startServer.js [hunk 15]
```diff
@@ -11,117 +11,147 @@
...
 type Options = {
   port?: number,
   method?: string,
   configDir?: string,
 };
 
 export default (async function startServer(options?: Options): Promise<void> {
-  if (!options || !options.configDir) {
+  if (!options) {
+    process.stderr.write('At least --configDir option is required!');
```
Comment (wincent) [inline packages/graphql-language-service-server/src/startServer.js]: Probably drop the "!"; you sound alarmed.

File: packages/graphql-language-service-server/src/startServer.js [hunk 16]
```diff
@@ -11,117 +11,147 @@
...
+    process.exit(1);
+    return;
+  }
+  if (!options.configDir) {
     process.stderr.write('--configDir is required!');
```
Comment (wincent) [inline packages/graphql-language-service-server/src/startServer.js]: Drop the "!". It's cleaner... (same below)?

File: packages/graphql-language-service-server/src/MessageProcessor.js (hunk 6 above)
Comment (asiandrummer) [inline packages/graphql-language-service-server/src/MessageProcessor.js]: Ah, just to throw an error when it's not there. It's an useful debugging "tool" for me

File: packages/graphql-language-service-server/src/startServer.js (hunk 14 above)
Comment (asiandrummer) [inline packages/graphql-language-service-server/src/startServer.js]: Yup, for flow - I'll go ask.

**Comment by asiandrummer** [comment]
I forgot to mention that this drops `getOutline` support - a subsequent PR will replace it with `textDocuments/documentSymbols`, which is LSP-compliant.

**Pull request merged**